### PR TITLE
Metallb FRR integration

### PR DESCRIPTION
**Pull request by fedepaol**
This PR connects the dots in FRR integration by collecting the remaining bits from https://github.com/metallb/metallb/pull/969 and:

- adding the reloader to the speaker image
- implementing the reload function
- adding a metallb_frr.yaml manifest to deploy metallb in frr experimental mode
- adding a CI lane that tests bgp in frr mode

### CODE REVIEW

**Comment by fedepaol** [comment]
Double checked the artifacts, at least the frr-ipv4 lane is really running in frr mode

File: internal/bgp/frr/config.go [hunk 1]
```diff
@@ -114,35 +117,39 @@ func writeConfig(config string, filename string) error {
...
-
-	// send HUP signal to FRR reloader
-	err = syscall.Kill(pidInt, syscall.SIGHUP)
-	if err != nil {
-		return err
-	}
-
-	return nil
+var reloadConfig = func(sf *singleflight.Group) error {
+	_, err, _ := sf.Do("reload",
```
Comment (oribon) [inline internal/bgp/frr/config.go:121]: nice!

File: internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden [hunk 2]
```diff
@@ -11,5 +11,10 @@ router bgp 100
   neighbor 10.2.2.254 remote-as 200
   neighbor 10.2.2.254 port 179
 
+  address-family ipv4 unicast
+    neighbor 10.2.2.254 activate
+  address-family ipv6 unicast
+    neighbor 10.2.2.254 activate
```
Comment (msherif1234) [inline internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden:17]: this isn't right to active v4 neigh from v6 AF ?

File: internal/bgp/frr/config.go [hunk 3]
```diff
@@ -35,7 +41,15 @@ router bgp {{.MyASN}}
...
 {{- end }}
 {{range $n := .Neighbors -}}
-{{range .Advertisements }}
+{{/* no bgp default ipv4-unicast prevents peering if no address families are defined. We declare an ipv4 one for the peer to make the pairing happen */}}
+{{- if eq (len .Advertisements) 0}}
+  address-family ipv4 unicast
+    neighbor {{$n.Addr}} activate
+  address-family ipv6 unicast
+    neighbor {{$n.Addr}} activate
+  exit-address-family
```
Comment (msherif1234) [inline internal/bgp/frr/config.go:50]: is this for testing ? we can't have bgp peer with having addresspool config if that happened it should have caught eariler or should have been caught if that isn't the case ?

File: internal/bgp/frr/config.go [hunk 4]
```diff
@@ -103,15 +117,39 @@ func writeConfig(config string, filename string) error {
...
+			}
+
+			// send HUP signal to FRR reloader
+			err = syscall.Kill(pidInt, syscall.SIGHUP)
+			if err != nil {
+				return nil, err
+			}
+			return nil, nil
+		})
+	return err
```
Comment (msherif1234) [inline internal/bgp/frr/config.go:146]: this is interesting package so the trick here is the sleep 2s u wanted to prevent doing back2back reload for 2sec window ?

File: internal/bgp/frr/config.go (hunk 3 above)
Comment (fedepaol) [reply internal/bgp/frr/config.go:50]: That's to mimic the same behaviour we have now.
Took me a while to find it. Without a service, frr won't establish a connection with the peer because of the `no bgp default ipv4-unicast`.

"By default, only the IPv4 unicast address family is announced to all neighbors. Using the ‘no bgp default ipv4-unicast’ configuration overrides this default so that all address families need to be enabled explicitly."
So without this, FRR will complain of no address families configured.

File: internal/bgp/frr/config.go (hunk 4 above)
Comment (fedepaol) [reply internal/bgp/frr/config.go:146]: Single flight squash all the calls that happen while one is in progress.
Adding the sleep will make the call last longer and will squash all the configuration requests into a single one (with the drawback of 2 seconds timeout). As a second thought, I can make the timeout even shorter since we care about squashing the calls caused by the same config change.

File: internal/bgp/frr/config.go (hunk 3 above)
Comment (msherif1234) [reply internal/bgp/frr/config.go:50]: that was my question I did noticed too if no service no bgp neighbors I couldn't answer if this is an ok thing or not why do we care of having bgp peering if we don't have any network to advertise , I don't know hence asked that queston

File: internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden (hunk 2 above)
Comment (fedepaol) [reply internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden:17]: I wasn't sure about this, but I think it's possible to advertise an ip6 family to a neighbour reachable via ipv4 (and viceversa) ?

File: internal/bgp/frr/config.go (hunk 4 above)
Comment (msherif1234) [reply internal/bgp/frr/config.go:146]: does it squash or block till on flight call completes ?

File: internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden (hunk 2 above)
Comment (fedepaol) [reply internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden:17]: It would be the same as https://github.com/metallb/metallb/pull/1014/files/b88845f28a5aa8ac430c75678d8414dec856ea14#diff-1d9a33c302812585b1b2d1897f84126492129d1f9415786a73dcfa26603686c9R53 (but we need to validate it).

Comment (msherif1234) [reply internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden:17]: I understand the config just I doubt if it will work as is i have seen some doc about how to get this to work 
```
Enable
Configure terminal
Route-Map IPV6 permit 10
Set ipv6 next-hop FE80::F816:3EFF:FE40:2985
Exit
Router bgp 1
Neighbor 2.0.0.2 remote-as 2
Address-family ipv6 unicast
Neighbor 2.0.0.2 activate
Neighbor 2.0.0.2 route-map IPV6 out
Network 2002:2::2/128
```
might be way more involved than u think

Comment (fedepaol) [reply internal/bgp/frr/testdata/TestSingleAdvertisementInvalidPrefix.golden:17]: This part is for enabling the session only (address-family without routes) to avoid the issue described here https://github.com/metallb/metallb/pull/1014#discussion_r740421183

I think we'll need to hammer our way through v6 / dual stack anyway, so I can either remove it or leave it as it is since it's not harming v4 e2e tests.

File: internal/bgp/frr/config.go (hunk 3 above)
Comment (fedepaol) [reply internal/bgp/frr/config.go:50]: I am not sure it's right or not, but the current implementation does that and it did not seem bad to replicate that.
Also, that would be the default frr behaviour if the `no bgp default ipv4-unicast` parameter (which is needed to allow both ipv4 and ipv6 to co-exist) wasn't added.

File: internal/bgp/frr/config.go (hunk 4 above)
Comment (fedepaol) [reply internal/bgp/frr/config.go:146]: The point about blocking makes total sense and I did not think about it. This will be likely to be called many subsequent times from the single goroutine, so the effect will be of delaying instead of really squashing them.
Let me put a goroutine in the middle so it will sort out the desired effect.

Comment (msherif1234) [reply internal/bgp/frr/config.go:146]: is there any possibility for this working async now that we could fire frr reload with hlaf baked config ? from what I have seen FRR will drop the entire config if one line is incorrect, just thinking loud here since we don't have any idea when the reload took effect debugging invalid config will be interesting

Comment (fedepaol) [reply internal/bgp/frr/config.go:146]: Configuration updates happen in discrete steps, and each step is supposed to provide a valid config (not meaning that we won't have bugs). What I am saying is that we won't produce a half baked config, we'll produce a config with less stuff that the caller expects:

What could happen here, is that the timeout falls in the middle (i.e. we are looping here https://github.com/metallb/metallb/blob/9cb1653a87673ec91971a60c873204e0752ff97b/speaker/bgp_controller.go#L172 and the threshold happens half way through https://github.com/metallb/metallb/blob/9cb1653a87673ec91971a60c873204e0752ff97b/speaker/bgp_controller.go#L269 )

With that, we would have less addresses announced, but the configuration (bugs aside) will still be valid.

In any case, I am torn about this because as it is now it's just introducing a delay and not working asynchronously and I don't want this to hold the integration to get in. Will remove the commit and rework it in a separate PR, using goroutines.

Comment (fedepaol) [reply internal/bgp/frr/config.go]: Removed the reloading optimization, will handle it in a separate PR

File: tasks.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 import sys
 import yaml
 import tempfile
+
```
Comment (russellb) [inline tasks.py:9]: Can you please re-do this commit without all of the unrelated formatting changes?  @cgoncalves

File: tasks.py [hunk 6]
```diff
@@ -251,9 +251,12 @@ def validate_kind_version():
     "node_img": "Optional node image to use for the kind cluster (e.g. kindest/node:v1.18.19)."
                 "The node image drives the kubernetes version used in kind.",
     "ip_family": "Optional ipfamily of the cluster."
-                 "Default: ipv4, supported families are 'ipv6' and 'dual'."
+                 "Default: ipv4, supported families are 'ipv6' and 'dual'.",
+    "bgp_type": "Type of BGP implementation to use."
+                "Supported: 'native' (default), 'frr'"
```
Comment (russellb) [inline tasks.py:256]: could have also been `-p bgp-native` and `-p bgp-frr` instead of a new option I guess, but this works too

Comment (fedepaol) [reply tasks.py:256]: We want to be able to differentiate, because CI now doesn't care about the protocol.
Using -p will add a pre-canned configuration which will interfere with the test execution.

Comment (russellb) [reply tasks.py:256]: good point, thanks

File: .circleci/config.yml [hunk 7]
```diff
@@ -27,18 +27,23 @@ jobs:
...
       - run: inv build
-      - run: inv dev-env -i  << parameters.ip-family >>
+      - run: inv dev-env -i  << parameters.ip-family >> -b  << parameters.bgp-type >>
       - run: |
-          if [ "<< parameters.ip-family >>" = "ipv4" ]; then SKIP="--skip IPV6"; fi
-          if [ "<< parameters.ip-family >>" = "dual" ]; then SKIP="--skip IPV6"; fi # TODO: until dual stack is supported, we can't run the ipv6 tests against the cluster
-          if [ "<< parameters.ip-family >>" = "ipv6" ]; then SKIP="--skip IPV4\|BGP"; fi # TODO: we are currently skipping BGP for ipv6, re enable when we enable the support
-          inv e2etest $SKIP -e /tmp/kind_logs
+          SKIP="none"
+          if [ "<< parameters.bgp-type >>" = "frr" ]; then SKIP="$SKIP\|metrics"; fi # TODO: enable metrics tests when implementing in FRR
```
Comment (cgoncalves) [inline .circleci/config.yml:41]: I'm not convinced with this SKIP approach for two reasons:
1. hard to human-parse and likely not scaling that well
2. it will run CI jobs that all they do is setup the environment and skip all tests, while reporting as "succeed" giving wrong idea that they work and are supported.

Can't we add invalid/unsupported topologies to the exclude list?

**Comment by cgoncalves** [comment]
Ugh, PR approved while reviewing it.

File: .circleci/config.yml (hunk 7 above)
Comment (fedepaol) [reply .circleci/config.yml:41]: Those are temporary. My expectation is that we are gonna remove the skips in a matter of days, and for the metric thing it does not make any sense to add a new combination to skip.
